File: backend/models/lineage.py
"""Data lineage API Pydantic models compatible with React Flow."""

from typing import Dict, List, Optional
from pydantic import BaseModel, Field


class LineageNode(BaseModel):
    """Lineage graph node representation."""

    id: str
    type: str = Field(..., description="Node type e.g. source, queue, engine, storage, sink")
    label: str
    status: Optional[str] = "HEALTHY"
    details: Optional[Dict[str, str]] = None


class LineageEdge(BaseModel):
    """Lineage graph directed edge representation."""

    id: str
    source: str
    target: str
    label: Optional[str] = None
    animated: bool = False


class LineageResponse(BaseModel):
    """React Flow compatible lineage graph response."""

    nodes: List[LineageNode]
    edges: List[LineageEdge]

File: frontend/src/types/lineage.ts
/**
 * Data Lineage TypeScript Types matching IceStream Telemetry Backend specifications.
 */

export type NodeStatus =
  | 'HEALTHY'
  | 'WARNING'
  | 'DEGRADED'
  | 'CRITICAL'
  | 'CIRCUIT_OPEN'
  | 'PAUSED'
  | 'UNKNOWN'
  | 'IDLE'
  | 'ACTIVE';

export type NodeType =
  | 'source'
  | 'queue'
  | 'engine'
  | 'storage'
  | 'quarantine'
  | 'dlq'
  | 'sink'
  | 'observability'
  | 'circuit_breaker'
  | 'remediation';

export interface LineageNodeData {
  label: string;
  type: string;
  status: NodeStatus;
  details?: Record<string, string>;
  selected?: boolean;
}

export interface ApiLineageNode {
  id: string;
  type: string;
  label: string;
  status?: string;
  details?: Record<string, string>;
}

export interface ApiLineageEdge {
  id: string;
  source: string;
  target: string;
  label?: string;
  animated?: boolean;
}

export interface ApiLineageResponse {
  nodes: ApiLineageNode[];
  edges: ApiLineageEdge[];
}

export interface PipelineSummary {
  pipeline_id: string;
  state: string;
  circuit_breaker_state: string;
  error_rate: number;
  open_incidents: number;
  last_updated?: string;
}
